File: security-platform/agents/frontend/src/redaction.ts
import { RedactionRule } from './types';

const DEFAULT_BLOCKLIST = [
  'password',
  'passwd',
  'pwd',
  'secret',
  'token',
  'api_key',
  'apikey',
  'auth',
  'credit_card',
  'card_number',
  'cvv',
  'ssn',
  'social_security',
  'pin',
  'passcode',
];

export class Redactor {
  private rules: RedactionRule[];
  private patterns: RegExp[];

  constructor(rules: RedactionRule[] = []) {
    this.rules = rules;
    this.patterns = DEFAULT_BLOCKLIST.map(field => 
      new RegExp(`\\b${field}\\b`, 'i')
    );
  }

  redact(obj: any): any {
    if (obj === null || obj === undefined) {
      return obj;
    }

    if (typeof obj === 'string') {
      return this.redactString(obj);
    }

    if (Array.isArray(obj)) {
      return obj.map(item => this.redact(item));
    }

    if (typeof obj === 'object') {
      const redacted: any = {};
      for (const [key, value] of Object.entries(obj)) {
        const redactedKey = this.shouldRedactKey(key) ? '[REDACTED]' : key;
        redacted[redactedKey] = this.redact(value);
      }
      return redacted;
    }

    return obj;
  }

  private shouldRedactKey(key: string): boolean {
    const lowerKey = key.toLowerCase();
    
    // Check default blocklist
    for (const pattern of this.patterns) {
      if (pattern.test(lowerKey)) {
        return true;
      }
    }

    // Check custom rules
    for (const rule of this.rules) {
      if (rule.field === '*' || rule.field === key) {
        try {
          const regex = new RegExp(rule.pattern, 'i');
          if (regex.test(key) || regex.test(lowerKey)) {
            return true;
          }
        } catch (e) {
          console.warn('Invalid regex pattern:', rule.pattern);
        }
      }
    }

    return false;
  }

  private redactString(str: string): string {
    // Check if string contains PII patterns
    for (const rule of this.rules) {
      if (!rule.field || rule.field === '*') {
        try {
          const regex = new RegExp(rule.pattern, 'gi');
          if (regex.test(str)) {
            return str.replace(regex, rule.replacement || '[REDACTED]');
          }
        } catch (e) {
          console.warn('Invalid regex pattern:', rule.pattern);
        }
      }
    }
    return str;
  }
}

export function redact(data: any, rules: RedactionRule[] = []): any {
  const redactor = new Redactor(rules);
  return redactor.redact(data);
}

File: security-platform/agents/python/src/agent/redaction.py
"""PII redaction implementation."""

import re
from typing import Any, List, Optional
from .types import RedactionRule

DEFAULT_BLOCKLIST = [
    "password",
    "passwd",
    "pwd",
    "secret",
    "token",
    "api_key",
    "apikey",
    "auth",
    "credit_card",
    "card_number",
    "cvv",
    "ssn",
    "social_security",
    "pin",
    "passcode",
]


class Redactor:
    """Redacts PII from data structures."""

    def __init__(self, rules: Optional[List[RedactionRule]] = None):
        """Initialize redactor with rules."""
        self.rules = rules or []
        self.patterns = [
            re.compile(rf"\b{field}\b", re.IGNORECASE) for field in DEFAULT_BLOCKLIST
        ]

    def redact(self, obj: Any) -> Any:
        """Recursively redact PII from object."""
        if obj is None:
            return obj

        if isinstance(obj, str):
            return self._redact_string(obj)

        if isinstance(obj, list):
            return [self.redact(item) for item in obj]

        if isinstance(obj, dict):
            redacted = {}
            for key, value in obj.items():
                redacted_key = (
                    "[REDACTED]" if self._should_redact_key(key) else key
                )
                redacted[redacted_key] = self.redact(value)
            return redacted

        return obj

    def _should_redact_key(self, key: str) -> bool:
        """Check if a key should be redacted."""
        lower_key = key.lower()

        # Check default blocklist
        for pattern in self.patterns:
            if pattern.search(lower_key):
                return True

        # Check custom rules
        for rule in self.rules:
            if rule.field == "*" or rule.field == key:
                try:
                    regex = re.compile(rule.pattern, re.IGNORECASE)
                    if regex.search(key) or regex.search(lower_key):
                        return True
                except re.error:
                    # Invalid regex, skip
                    pass

        return False

    def _redact_string(self, s: str) -> str:
        """Redact PII from string."""
        for rule in self.rules:
            if not rule.field or rule.field == "*":
                try:
                    regex = re.compile(rule.pattern, re.IGNORECASE)
                    if regex.search(s):
                        return regex.sub(rule.replacement, s)
                except re.error:
                    # Invalid regex, skip
                    pass
        return s
